File: client/src/api/ssd-cache.ts
import { buildApiUrl } from '../lib/api';
import type { RaidActionResponse } from './raid';

export type CacheMode = 'writethrough' | 'writeback' | 'writearound' | 'none';

export interface CacheStatus {
  array_name: string;
  cache_device: string;
  bcache_device: string | null;
  mode: CacheMode;
  state: string;
  hit_rate_percent: number | null;
  dirty_data_bytes: number;
  cache_size_bytes: number;
  cache_used_bytes: number;
  sequential_cutoff_bytes: number;
}

export interface CacheAttachPayload {
  array: string;
  cache_device: string;
  mode?: CacheMode;
}

export interface CacheDetachPayload {
  array: string;
  force?: boolean;
}

export interface CacheConfigurePayload {
  array: string;
  mode?: CacheMode;
  sequential_cutoff_bytes?: number;
}

export interface ExternalBitmapPayload {
  array: string;
  ssd_partition: string;
}

const getToken = (): string => {
  const token = localStorage.getItem('token');
  if (!token) {
    throw new Error('Keine aktive Sitzung – bitte erneut anmelden.');
  }
  return token;
};

const parseResponse = async <T>(response: Response): Promise<T> => {
  const text = await response.text();
  let payload: any = null;

  if (text) {
    try {
      payload = JSON.parse(text);
    } catch {
      payload = null;
    }
  }

  if (!response.ok) {
    const detail =
      (payload && (payload.detail || payload.error || payload.message)) ||
      `Request failed with status ${response.status}`;
    throw new Error(typeof detail === 'string' ? detail : 'SSD cache request failed');
  }

  return payload as T;
};

export const getCacheStatuses = async (): Promise<CacheStatus[]> => {
  const token = getToken();
  const response = await fetch(buildApiUrl('/api/system/raid/cache/status'), {
    headers: { Authorization: `Bearer ${token}` },
  });
  return parseResponse<CacheStatus[]>(response);
};

export const getCacheStatus = async (array: string): Promise<CacheStatus> => {
  const token = getToken();
  const response = await fetch(buildApiUrl(`/api/system/raid/cache/status/${encodeURIComponent(array)}`), {
    headers: { Authorization: `Bearer ${token}` },
  });
  return parseResponse<CacheStatus>(response);
};

export const attachCache = async (payload: CacheAttachPayload): Promise<RaidActionResponse> => {
  const token = getToken();
  const response = await fetch(buildApiUrl('/api/system/raid/cache/attach'), {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(payload),
  });
  return parseResponse<RaidActionResponse>(response);
};

export const detachCache = async (payload: CacheDetachPayload): Promise<RaidActionResponse> => {
  const token = getToken();
  const response = await fetch(buildApiUrl('/api/system/raid/cache/detach'), {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(payload),
  });
  return parseResponse<RaidActionResponse>(response);
};

export const configureCache = async (payload: CacheConfigurePayload): Promise<RaidActionResponse> => {
  const token = getToken();
  const response = await fetch(buildApiUrl('/api/system/raid/cache/configure'), {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(payload),
  });
  return parseResponse<RaidActionResponse>(response);
};

export const setExternalBitmap = async (payload: ExternalBitmapPayload): Promise<RaidActionResponse> => {
  const token = getToken();
  const response = await fetch(buildApiUrl('/api/system/raid/cache/external-bitmap'), {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(payload),
  });
  return parseResponse<RaidActionResponse>(response);
};

File: backend/app/schemas/ssd_cache.py
"""Pydantic schemas for SSD cache (bcache) management."""

from enum import Enum

from pydantic import BaseModel


class CacheMode(str, Enum):
    WRITETHROUGH = "writethrough"  # Safe: no data loss on SSD failure
    WRITEBACK = "writeback"        # Fast: writes cached on SSD
    WRITEAROUND = "writearound"    # Only reads cached
    NONE = "none"


class CacheStatus(BaseModel):
    """Status of an SSD cache attached to a RAID array."""
    array_name: str                          # e.g. "md0"
    cache_device: str                        # e.g. "nvme1n1p1"
    bcache_device: str | None = None         # e.g. "bcache0"
    mode: CacheMode
    state: str                               # running, detaching, error, idle
    hit_rate_percent: float | None = None
    dirty_data_bytes: int = 0                # Only relevant in writeback mode
    cache_size_bytes: int = 0
    cache_used_bytes: int = 0
    sequential_cutoff_bytes: int = 4 * 1024 * 1024  # 4 MiB default


class CacheAttachRequest(BaseModel):
    """Request to attach an SSD cache to a RAID array."""
    array: str
    cache_device: str
    mode: CacheMode = CacheMode.WRITETHROUGH


class CacheDetachRequest(BaseModel):
    """Request to detach an SSD cache from a RAID array."""
    array: str
    force: bool = False


class CacheConfigureRequest(BaseModel):
    """Request to change cache parameters."""
    array: str
    mode: CacheMode | None = None
    sequential_cutoff_bytes: int | None = None


class ExternalBitmapRequest(BaseModel):
    """Request to move RAID bitmap to SSD."""
    array: str
    ssd_partition: str
